File: INT102/7week/cpp/src/counting_sort.cc
#include "counting_sort.h"

#include <math.h>

#include <iostream>
#define DEBUG

void countingSort() {
    int input[5] = {4, 3, 3, 1, 1};
    int min = input[0];
    int max = input[0];

    // create the cnt array
    for (int i = 0; i < 5; i++) {
        min = std::min(input[i], min);
        max = std::max(input[i], max);
    }

    const int cnt_size = max - min + 1;  // size of the counting array
    int cnt[cnt_size];
    for (int i = 0; i < cnt_size; i++) {
        cnt[i] = 0;
    }

    for (int i = 0; i < 5; i++) {
        cnt[input[i] - min]++;
    }

    int temp_idx = 0;
    for (int i = 0; i < cnt_size; i++) {
        for (int j = 0; j < cnt[i]; j++) {
            input[temp_idx++] = i + min;  // i + min to get the original value
        }
    }

#ifdef DEBUG
    std::string sb = "";
    for (int i = 0; i < 5; i++) {
        sb += std::string(" ") + std::to_string(input[i]);
    }
    std::cout << sb << std::endl;
#endif  // DEBUG

    return;
}

File: INT102/7week/sortingshit/countingSort.cs
using System;

namespace ns_counting_sort
{
    public class SHIT
    {
        public SHIT()
        {
            const int input_size = 10;
            int[] input = new int[input_size] { 9, 9, 8, 8, 4, 4, 5, 1, 2, 2 };

            // find the min and the max
            int min = input.Min(),
                max = input.Max();

            // create a counting array
            int[] cnt = new int[max - min + 1];

            // counting shit
            foreach (var shit in input)
                cnt[shit - min]++;

            // rearrange the input array
            int[] output = new int[input_size];
            int temp_idx = 0;
            for (int i = 0; i < max - min + 1; i++)
            {
                if (cnt[i] == 0)
                {
                    continue;
                }
                else
                {
                    for (int j = 0; j < cnt[i]; j++)
                    {
                        output[temp_idx] = i + min;
                        temp_idx++;
                    }
                }
            }
            Console.WriteLine(String.Join(" ", output));
        }
    }
}
